File: reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        stack<int> stack0;
        
        while (head){
            stack0.push(head->val);
            head = head->next;
        }
        
        ListNode dummy_head(0);
        ListNode* tail = &dummy_head;
        
        while (!stack0.empty()){
            tail->next = new ListNode(stack0.top());
            stack0.pop();
            tail = tail->next;
        }
        return dummy_head.next;
    }
};

File: reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        stack = []
        while head:
            stack.append(head.val)
            head = head.next
        
        dummy_head = tail = ListNode(0)
        
        while stack:
            tail.next = ListNode(stack.pop(-1))
            tail = tail.next
        return dummy_head.next